[exporter/bmchelix] enhance metric name disambiguation using attributes

NassimBtk opened the pull request:
This pull request enhances the `bmchelixexporter` component by improving metric name disambiguation using attributes. Key changes include adding a mechanism to enrich metric names with identifying attributes, ensuring compatibility with BMC Helix Operations Management, and introducing tests to validate the new functionality.

### Enhancements to Metric Name Disambiguation:
* Added a new function, `enrichMetricNamesWithAttributes`, to append distinguishing attributes to metric names when attributes have more than one distinct value across metrics with the same `entityId` and `metricName`. This ensures unique and meaningful metric names in the BMC Helix Operations Management payload. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`)
* Introduced a `coreAttributes` map to specify attributes that should be ignored when building metric name suffixes. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`)

### Compatibility Improvements:
* Updated the `updateEntityInformation` method to sanitize `entityName` by trimming and removing colons for compatibility with BMC Helix Operations Management. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`)

### Testing:
* Added a new test, `TestEnrichMetricNamesWithAttributes`, to validate the behavior of the `enrichMetricNamesWithAttributes` function under various scenarios, ensuring robustness. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer_test.go`)

### Documentation:
* Created a changelog entry to document the enhancement for release notes. (`.chloggen/bmchelixexporter_enhance_metric_name_disambiguation_with_attributes.yaml`)

#### Link to tracking issue
Fixes #41303

NassimBtk committed:
[exporter/bmchelix] enhance metric name disambiguation using attributes

- Add `coreAttributes` to exclude specific labels from name suffixes
- Implement `enrichMetricNamesWithAttributes` for unique metric names
- Modify `ProduceHelixPayload` to enrich metric names before returning
- Ensure compatibility by sanitizing `entityName` in label processing
- Add unit tests for `enrichMetricNamesWithAttributes` with various
cases
- Update test cases to validate enriched metric name generation logic

NassimBtk committed:
[exporter/bmchelix] enhance metric name disambiguation using attributes

- Code format

bertysentry reviewed (approved):
Very nice!

It will be useful, optionally, to keep the original metrics so they are pushed to BMC Helix's VictoriaMetrics and can be queried in "regular" PromQL

NassimBtk committed:
[exporter/bmchelix] enhance metric name disambiguation using attributes

- Removed unused `sort` import from `metrics_producer.go`.
- Added filtering for metrics with empty `entityTypeId` labels.
- Refactored `enrichMetricNamesWithAttributes` to preserve original
metrics.
- Updated test cases to validate enriched metric labels instead of
names.
- Adjusted test data to include detailed label expectations.
- Ensured backward compatibility with existing metric processing logic.

bertysentry commented on exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go:
```diff
@@ -266,3 +294,66 @@ func extractResourceAttributes(resource pcommon.Resource) map[string]string {
+				identifyingKeys = append(identifyingKeys, k)
+			}
+		}
+		sort.Strings(identifyingKeys) // Ensure attributes ordering
```
Thank you for running the benchmark! 😅
